File: WalletApp.cpp
#include <iostream>
#include <vector>

class WalletUser {
public:
    std::string username;
    std::string password;
    double balance;
    std::vector<std::string> transactions;

    WalletUser(std::string username, std::string password) {
        this->username = username;
        this->password = password;
        this->balance = 0.0;
    }
};

WalletUser* current_user = nullptr;

// Déclaration des fonctions
void show_main_menu();
void display_balance();
void show_transaction_history();
void deposit_money();
void withdraw_money();
void transfer_money();

void create_user() {
    std::cout << "Entrez un nom d'utilisateur : ";
    std::string username;
    std::cin >> username;

    std::cout << "Entrez un mot de passe : ";
    std::string password;
    std::cin >> password;

    current_user = new WalletUser(username, password);
    std::cout << "Compte créé avec succès !" << std::endl;
}

void login_user() {
    std::cout << "Entrez le nom d'utilisateur : ";
    std::string username;
    std::cin >> username;

    std::cout << "Entrez le mot de passe : ";
    std::string password;
    std::cin >> password;

    if (current_user != nullptr && username == current_user->username && password == current_user->password) {
        std::cout << "Connexion réussie !" << std::endl;
        show_main_menu();
    } else {
        std::cout << "Identifiants incorrects. Veuillez réessayer." << std::endl;
    }
}

void show_main_menu() {
    while (true) {
        std::cout << "\nMenu Principal :" << std::endl;
        std::cout << "1. Afficher le Solde" << std::endl;
        std::cout << "2. Historique des Transactions" << std::endl;
        std::cout << "3. Déposer de l'Argent" << std::endl;
        std::cout << "4. Retirer de l'Argent" << std::endl;
        std::cout << "5. Transférer de l'Argent" << std::endl;
        std::cout << "6. Se Déconnecter" << std::endl;

        int choice;
        std::cout << "Entrez votre choix : ";
        std::cin >> choice;

        switch (choice) {
            case 1:
                display_balance();
                break;
            case 2:
                show_transaction_history();
                break;
            case 3:
                deposit_money();
                break;
            case 4:
                withdraw_money();
                break;
            case 5:
                transfer_money();
                break;
            case 6:
                std::cout << "Déconnexion..." << std::endl;
                delete current_user;
                current_user = nullptr;
                return;
            default:
                std::cout << "Choix invalide. Veuillez réessayer." << std::endl;
        }
    }
}

void display_balance() {
    std::cout << "Solde actuel : $" << current_user->balance << std::endl;
}

void show_transaction_history() {
    std::cout << "Historique des transactions :" << std::endl;
    for (const std::string& transaction : current_user->transactions) {
        std::cout << transaction << std::endl;
    }
}

void deposit_money() {
    double amount;
    std::cout << "Entrez le montant à déposer : ";
    std::cin >> amount;

    current_user->balance += amount;
    current_user->transactions.push_back("Dépôt de $" + std::to_string(amount));
    std::cout << "Dépôt réussi !" << std::endl;
}

void withdraw_money() {
    double amount;
    std::cout << "Entrez le montant à retirer : ";
    std::cin >> amount;

    if (current_user->balance >= amount) {
        current_user->balance -= amount;
        current_user->transactions.push_back("Retrait de $" + std::to_string(amount));
        std::cout << "Retrait réussi !" << std::endl;
    } else {
        std::cout << "Solde insuffisant." << std::endl;
    }
}

void transfer_money() {
    std::string recipient;
    double amount;

    std::cout << "Entrez le nom d'utilisateur du destinataire : ";
    std::cin >> recipient;

    std::cout << "Entrez le montant à transférer : ";
    std::cin >> amount;

    if (current_user->balance >= amount) {
        current_user->balance -= amount;
        current_user->transactions.push_back("Transfert de $" + std::to_string(amount) + " à " + recipient);
        std::cout << "Transfert réussi !" << std::endl;
    } else {
        std::cout << "Solde insuffisant." << std::endl;
    }
}

int main() {
    while (true) {
        std::cout << "1. Créer un Compte" << std::endl;
        std::cout << "2. Se Connecter" << std::endl;
        std::cout << "3. Quitter" << std::endl;

        int choice;
        std::cout << "Entrez votre choix : ";
        std::cin >> choice;

        switch (choice) {
            case 1:
                create_user();
                break;
            case 2:
                login_user();
                break;
            case 3:
                std::cout << "Au revoir !" << std::endl;
                return 0;
            default:
                std::cout << "Choix invalide. Veuillez réessayer." << std::endl;
        }
    }

    return 0;
}

File: WalletApp.py
class WalletUser:
    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.balance = 0.0
        self.transactions = []

def create_user():
    print("Entrez un nom d'utilisateur : ")
    username = input().strip()

    print("Entrez un mot de passe : ")
    password = input().strip()

    global current_user
    current_user = WalletUser(username, password)
    print("Compte créé avec succès !")

def login_user():
    print("Entrez le nom d'utilisateur : ")
    username = input().strip()

    print("Entrez le mot de passe : ")
    password = input().strip()

    global current_user
    if username == current_user.username and password == current_user.password:
        print("Connexion réussie !")
        show_main_menu()
    else:
        print("Identifiants incorrects. Veuillez réessayer.")

def show_main_menu():
    while True:
        print("\nMenu Principal :")
        print("1. Afficher le Solde")
        print("2. Historique des Transactions")
        print("3. Déposer de l'Argent")
        print("4. Retirer de l'Argent")
        print("5. Transférer de l'Argent")
        print("6. Se Déconnecter")

        choice = input("Entrez votre choix : ").strip()

        if choice == "1":
            display_balance()
        elif choice == "2":
            show_transaction_history()
        elif choice == "3":
            deposit_money()
        elif choice == "4":
            withdraw_money()
        elif choice == "5":
            transfer_money()
        elif choice == "6":
            print("Déconnexion...")
            global current_user
            current_user = None
            break
        else:
            print("Choix invalide. Veuillez réessayer.")

def display_balance():
    print(f"Solde actuel : ${current_user.balance:.2f}")

def show_transaction_history():
    print("Historique des transactions :")
    for transaction in current_user.transactions:
        print(transaction)

def deposit_money():
    amount = float(input("Entrez le montant à déposer : "))
    current_user.balance += amount
    current_user.transactions.append(f"Dépôt de ${amount:.2f}")
    print("Dépôt réussi !")

def withdraw_money():
    amount = float(input("Entrez le montant à retirer : "))
    if current_user.balance >= amount:
        current_user.balance -= amount
        current_user.transactions.append(f"Retrait de ${amount:.2f}")
        print("Retrait réussi !")
    else:
        print("Solde insuffisant.")

def transfer_money():
    recipient = input("Entrez le nom d'utilisateur du destinataire : ").strip()
    amount = float(input("Entrez le montant à transférer : "))
    if current_user.balance >= amount:
        current_user.balance -= amount
        current_user.transactions.append(f"Transfert de ${amount:.2f} à {recipient}")
        print("Transfert réussi !")
    else:
        print("Solde insuffisant.")

while True:
    print("1. Créer un Compte")
    print("2. Se Connecter")
    print("3. Quitter")

    choice = input("Entrez votre choix : ").strip()

    if choice == "1":
        create_user()
    elif choice == "2":
        login_user()
    elif choice == "3":
        print("Au revoir !")
        break
    else:
        print("Choix invalide. Veuillez réessayer.")
